runtime/lua/vim/treesitter/query.lua

```diff
@@ -382,6 +382,36 @@ local predicate_handlers = {
 
     return string_set[node_text]
   end,
+
+  ['has-ancestor?'] = function(match, _, _, predicate)
+    local node = match[predicate[2]]
+    if not node then
+      return true
+    end
+
+    local ancestor_types = { unpack(predicate, 3) }
+    node = node:parent()
+    while node do
+      if vim.tbl_contains(ancestor_types, node:type()) then
```

Comment:
Just passing by, but can't this be made a lot faster by turning the `ancestor_type` into a map ?
Something along those line:
```lua
for _, ty in ipairs(ancestor_types) do
  map[ty] = true
end
```
And then use it directly here:
```suggestion
      if map[node:type()] then
```

Reply:
Depends on how large the list is, possibly. (In any case, this _is_ a list, so `list_contains` should be used.)